lib/shared/puppet_forge/unpacker.rb

```diff
@@ -0,0 +1,78 @@
+require 'pathname'
+require 'tmpdir'
+require 'shared/puppet_forge/error'
+require 'shared/puppet_forge/tar'
+
+module PuppetForge
+  class Unpacker
+    # Unpack a tar file into a specified directory
+    #
+    # @param filename [String] the file to unpack
+    # @param target [String] the target directory to unpack into
+    def self.unpack(filename, target, tmpdir)
+      inst = self.new(filename, target, tmpdir)
+      inst.unpack
+      inst.move_into(target)
+    end
+
+    # Set the owner/group of the target directory to those of the source
+    #
+    # @param source [Pathname] source of the permissions
+    # @param target [Pathname] target of the permissions change
+    # @param microsoft_windows boolean flag for whether this system is windows or not
+    def self.harmonize_ownership(source, target, microsoft_windows = false)
+      if !microsoft_windows
+        FileUtils.chown_R(source.stat.uid, source.stat.gid, target)
+      end
+    end
+
+    # @param filename [String] the file to unpack
+    # @param target [String] the target directory to unpack into
+    def initialize(filename, target, tmpdir)
+      @filename = Pathname.new(filename)
+      @target = target
+      @tmpdir = tmpdir
+    end
+
+    # @api private
+    def unpack
+      begin
+        PuppetForge::Tar.instance.unpack(@filename.to_s, @tmpdir)
+      rescue PuppetForge::ExecutionFailure => e
+        raise RuntimeError, "Could not extract contents of module archive: #{e.message}"
+      end
+    end
+
+    # @api private
+    def move_into(dir)
```

Comment:
This is problematic because the directory that will be moved is in the format `username-modulename-version`, which means that we can't move this from the temp location to the final location without stuffing another rename into the mix.

```
    52: def move_into(dir)
    53:   dir = Pathname.new(dir)
    54:   dir.rmtree if dir.exist?
 => 55:   require 'pry'; binding.pry
    56:   FileUtils.mv(root_dir, dir)
    57: ensure
    58:   FileUtils.rmtree(tmpdir)
    59: end

[1] pry(#<PuppetForge::Unpacker>)> dir
=> #<Pathname:/tmp/d20150513-13668-10i1lta>
[2] pry(#<PuppetForge::Unpacker>)> root_dir
=> #<Pathname:/tmp/tmp-unpacker20150513-13668-1x49wtg/branan-eight_hundred-8.0.0>
```